parlai/core/utils.py

```diff
@@ -745,17 +750,31 @@ def str_segment(self, text, dict_agent, max_length):
         V = len(freqs)
 
         logNV = math.log(N + V)
+        max_heap = []
+
+        if not max_length:
+            max_length = len(text)
 
-        @lru_cache(maxsize=None)
+        @lru_cache(maxsize=16)
         def segment(text):
             # Return a list of words that is the best segmentation of text.
             if not text:
                 return []
-            candidates = (
+            candidates = [
                 [first] + segment(rem)
                 for first, rem in splits(text, max_length)
-            )
-            return max(candidates, key=score)
+            ]
+
+            nonlocal max_heap
+            max_heap = []
+
+            for c in candidates:
+                cand_score = (score(c), c)  # tuple of (score, candidate)
+                max_heap.append(cand_score)
```

Comment:
if we have any reason to believe the memory usage here might get heavy (large number of candidates), then it would be better to use the priority queue from the ir baseline class as it never stores more than k elements at once (instead of strong them all and then popping the top k) but I bet that won't come up here, ha

Reply:
Yeah it's weird, I learned today that there is no fixed sized heap in python